lambda-buffers-compiler/src/LambdaBuffers/Compiler/TypeClass/Solve.hs

```diff
@@ -0,0 +1,131 @@
+import Control.Monad.Writer.Strict (WriterT, execWriterT)
+import Data.Foldable (traverse_)
+import Data.List (foldl')
+import Data.Set qualified as S
+import Data.Text (Text)
+
+{- Pattern/Template/Unification variable  substitution.
+   Given a string that represents a variable name,
+   and a type to instantiate variables with that name to,
+   performs the instantiation
+-}
+subV :: Text -> Pat -> Pat -> Pat
+subV varNm t = \case
+  var@(VarP v) -> if v == varNm then t else var
+  x :* xs -> subV varNm t x :* subV varNm t xs
+  l := x -> subV varNm t l := subV varNm t x
+  ProdP xs -> ProdP (subV varNm t xs)
+  RecP xs -> RecP (subV varNm t xs)
+  SumP xs -> SumP (subV varNm t xs)
+  AppP t1 t2 -> AppP (subV varNm t t1) (subV varNm t t2)
+  RefP n x -> RefP (subV varNm t n) (subV varNm t x)
+  DecP a b c -> DecP (subV varNm t a) (subV varNm t b) (subV varNm t c)
+  other -> other
+
+{- Performs substitution on an entire instance (the first argument) given the
+   concrete types from a Pat (the second argument).
+   Note that ONLY PatVars which occur in the Instance *HEAD* are replaced, though they
+   are replaced in the instance superclasses as well (if they occur there).
+-}
+subst :: Rule -> Pat -> Rule
+subst cst@(C _ t :<= _) ty = mapPat (go (getSubs t ty)) cst
+  where
+    go :: [(Text, Pat)] -> Pat -> Pat
+    go subs tty =
+      let noflip p1 p2 = uncurry subV p2 p1
+       in foldl' noflip tty subs
+
+{- Given two patterns (which are hopefully structurally similar), gather a list of all substitutions
+   from the PatVars in the first argument to the concrete types (hopefully!) in the second argument
+-}
+getSubs :: Pat -> Pat -> [(Text, Pat)] -- should be a set, whatever
+getSubs (VarP s) t = [(s, t)]
+getSubs (x :* xs) (x' :* xs') = getSubs x x' <> getSubs xs xs'
+getSubs (l := t) (l' := t') = getSubs l l' <> getSubs t t'
+getSubs (ProdP xs) (ProdP xs') = getSubs xs xs'
+getSubs (RecP xs) (RecP xs') = getSubs xs xs'
+getSubs (SumP xs) (SumP xs') = getSubs xs xs'
+getSubs (AppP t1 t2) (AppP t1' t2') = getSubs t1 t1' <> getSubs t2 t2'
+getSubs (RefP n t) (RefP n' t') = getSubs n n' <> getSubs t t'
+getSubs (DecP a b c) (DecP a' b' c') = getSubs a a' <> getSubs b b' <> getSubs c c'
+getSubs _ _ = []
+
+-- NoMatch isn't fatal but OverlappingMatches is (i.e. we need to stop when we encounter it)
+data MatchError
+  = NoMatch
+  | OverlappingMatches [Rule]
+
+-- for SolveM, since we catch NoMatch
+data Overlap = Overlap Constraint [Rule]
+  deriving stock (Show, Eq)
+
+selectMatchingInstance :: Pat -> Class -> [Rule] -> Either MatchError Rule
+selectMatchingInstance p c rs = case filter matchPatAndClass rs of
+  [] -> Left NoMatch
+  [r] -> Right r
+  overlaps -> Left $ OverlappingMatches overlaps
+  where
+    matchPatAndClass :: Rule -> Bool
+    matchPatAndClass r =
+      ruleHeadClass r == c
+        && ruleHeadPat r
+        `matches` p
+
+type SolveM = ReaderT [Rule] (WriterT (S.Set Constraint) (Either Overlap))
```

Comment:
I don't think `Map RuleHead Rule` is actually useful here? We're not doing straightforward lookups, we're filtering the List/Set/whatever of `Rule`s to find a match. A map doesn't really get us anything that we don't have w/ a list. 

From glancing at your protocompat PR it doesn't appear that `SolveError` is implemented ATM. That's fine, it might be easier to write the protos for all of the API errors after/inside of the next PR (since that's where the basic conditions for instance validity are)